playkit/src/main/java/com/kaltura/playkit/ads/AdEnabledPlayerController.java

```diff
@@ -1,24 +1,49 @@
 package com.kaltura.playkit.ads;
 
+import android.support.annotation.NonNull;
+
+import com.kaltura.playkit.PKEvent;
 import com.kaltura.playkit.PKLog;
+import com.kaltura.playkit.PlayerConfig;
 import com.kaltura.playkit.PlayerDecorator;
+import com.kaltura.playkit.plugins.ads.AdEvent;
 import com.kaltura.playkit.plugins.ads.AdsProvider;
 
+import static com.kaltura.playkit.plugins.ads.AdError.Type.*;
+
 /**
  * Created by gilad.nadav on 20/11/2016.
  */
 
-public class AdEnabledPlayerController extends PlayerDecorator implements AdController {
+public class AdEnabledPlayerController extends PlayerDecorator implements AdController, PKAdEventListener {
 
     private static final PKLog log = PKLog.get("AdEnablController");
 
-    AdsProvider adsProvider;
+    private AdsProvider adsProvider;
+    private boolean isPlayerPrepared;
+    private PlayerConfig.Media mediaConfig;
+
     public AdEnabledPlayerController(AdsProvider adsProvider) {
         log.d("Init AdEnabledPlayerController");
         this.adsProvider = adsProvider;
     }
 
     @Override
+    public void prepare(@NonNull final PlayerConfig.Media mediaConfig) {
+        this.mediaConfig = mediaConfig;
+
+        isPlayerPrepared = false;
+        if (adsProvider != null) {
+            if (adsProvider.isAdRequested()) {
+                super.prepare(mediaConfig);
+                isPlayerPrepared = true;
+            } else {
+                adsProvider.setAdLoadedListener(this);
+            }
+        }
+    }
+
+    @Override
```

Comment:
Didn't we agree that getDuration() always returns the content duration? I remember something like that.